### ISSUE TITLE
STDIN/STDOUT server

### ISSUE THREAD

**Issue by parisk** [paritytech/jsonrpc#194]
Hi there,

I would like to ask a question about this project.

Is there a way to set up a "server" that reads JSON RPC requests from STDIN and prints the output to STDOUT?

I have read part of the source code, but I am not quite sure yet how it works.

**Disclaimer**: I am also totally new to Rust, so this could be total newbie question 😄.

**Comment by tomusdrw** [comment]
Hey!
Currently there is no such server implemented, but the implementation should be pretty straight forward.

You can however use `jsonrpc-tcp-server` and `nc` tool:
```
$ cargo run --example tcp -p jsonrpc-tcp-server
$ echo '{"jsonrpc":"2.0","id":1,"method":"say_hello","params":[]}' | nc -v localhost 3030
Connection to localhost 3030 port [tcp/*] succeeded!
{"jsonrpc":"2.0","result":"hello","id":1}
```

**Comment by bemeurer** [comment]
I'll be taking this on :) 

@tomusdrw, any tips?

**Comment by tomusdrw** [comment]
@bemeurer 
Perhaps consider those libraries:

https://github.com/paulkernfeld/tokio-stdin
https://github.com/paulkernfeld/tokio-stdout

Author mentions that these are not production ready, but might be a good opportunity to contribute to that projects as well. That said I don't think that stdin/stdout jsonrpc server would ever be useful in production environment, so we don't have to worry that much.

**Comment by bemeurer** [comment]
@tomusdrw I'm starting work on this right now. If it's not too much to ask, and mostly because I haven't done anything with the jsonrpc library before, could you give me some mentoring on what needs to be done?
What I mean by that is some kind of either roadmap or a north on where to start.

**Comment by tomusdrw** [comment]
Hi @bemeurer, personally I like to start from an example or a test:
https://github.com/paritytech/jsonrpc/blob/master/ws/examples/ws.rs#L5
(you don't need to start with super complicated `ServerBuilder` at the beginning)

Later it would be just a matter of calling `IoHandler::handle_request` method: https://github.com/paritytech/jsonrpc/blob/master/core/src/io.rs#L169

So you only need to get a new-line delimited string from `tokio-stdin` and then pass the response to `tokio-stdout` when it's ready. Could probably be done nicely using `Stream` and `Sink` traits initially.

At the very beginning I would skip:
1. Metadata,
2. Extractors
3. Sessions & PubSub

just to make the simplest example work and later add those features in subsequent commits/PRs.
Hope it helps. Feel free to reach me out on Gitter as well for more interactive help.

**Comment by cmichi** [comment]
Hi there, I quickly wanted to ask if there is still any work done on this or if I could otherwise take it over?

I'm in the same situation of wanting to dive into Rust and this would be an ideal starting point for me.

**Comment by tomusdrw** [comment]
@cmichi I don't think anyone is on it currently, feel free to start working on this. Let me know if you have any issues on the way.

**Comment by cmichi** [comment]
Can be closed because #299 was merged.
